### PR TITLE
VMI configuration test: fix disk cache modes testing

### PR DESCRIPTION
**Pull request by enp0s3**
Suggested solution is to skip testing the cache modes 
for block volume and host path.

Signed-off-by: Igor Bezukh <ibezukh@redhat.com>

**What this PR does / why we need it**:
Block volume and host-path PVs and the corresponding PVCs
are being created as part of the test suite initialization.
However, these PVCs cannot be assumed to natively support Direct I/O
because the underlying FS that backs the PV may not support it.
One such example was the KIND provider that mounts local PVs on tmpfs.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Review by rmohr — CHANGES_REQUESTED** [review]
I am all for adjusting to the environment, so that we can run in a broad set of clusters, bit I think that goes too far. As far as I understand it, we want to test that `none` gets set. If we would change the test in this way, we can not test as exactly. I would suggest skipping the test in such an environment. At a certain point skipping is better than making tests vague.

**Comment by ezrasilvera** [comment]
The problem of the change is that it make the test not specific which bit the purpose of functional test. You want to take a known situation X and make sure the result was Y.
In this case take a FS that supports O_DIRECT -->  Default set the NONE, FS that doesn't support O_DIRECT --> writethrough. 
With this we test the  overall path - for example the functions that discover whether the FS supports O_DIRECT or not. 
By changing to "BeElementOf(cacheWritethrough, cacheNone)"  you basically eliminate the test completely as the input and results are not explicitly defined a-priory

**Comment by enp0s3** [comment]
@rmohr @ezrasilvera Thank you for bringing that up. Yep I agree with both of you. Need to think whether it is possible to know ahead-of-time if the underlying file system supports Direct I/O. I could suggest to add some custom annotations of the PVC for example, throwing the responsibility on the administrator. But maybe will come up with an idea to determine it automatically.

**Comment by ezrasilvera** [comment]
> @rmohr @ezrasilvera Thank you for bringing that up. Yep I agree with both of you. Need to think whether it is possible to know ahead-of-time if the underlying file system supports Direct I/O. I could suggest to add some custom annotations of the PVC for example, throwing the responsibility on the administrator. But maybe will come up with an idea to determine it automatically.

I don't think you need to check it in advance. This is what the checkDirectIOFlag() function is doing as part of setting the cachemode which is what this test does.
You need to create a known PVC, if you can't we need to re-think the whole test.  

This is a general issue not related just to this test. It is related to all tests that depend on type of resources that might change in different environments/providers 
I think we can expect the number of providers and envs will increase over time and we probably don't want to start if/else all over the tests code. 
For example in this case - if we can't create a block volume with a known O_DIRECT support I'm not sure if we should test it at all.

**Comment by enp0s3** [comment]
@ezrasilvera @rmohr The block volume PV is mounted at **/mnt/local-storage** and the HostPath PV is mounted at **/tmp?hostImages** correct me if I'm wrong but we cannot guarantee that the **/mnt** and **/tmp** are mounted to a FS that supports O_DIRECT. I can suggest 2 possible solutions:
1. In our documentation, update the tests section with the relevant information about pre-conditions that should be met regarding the mount points and the direct I/O support.
2. Remove the part in this particular test that validates the cache mode for host-path and block PVCs.

WDYT?

**Comment by xpivarc** [comment]
> @ezrasilvera @rmohr The block volume PV is mounted at **/mnt/local-storage** and the HostPath PV is mounted at **/tmp/hostImages** correct me if I'm wrong but we cannot guarantee that the **/mnt** and **/tmp** are mounted to a FS that supports O_DIRECT. I can suggest 2 possible solutions:
> 
>     1. In our documentation, update the tests section with the relevant information about pre-conditions that should be met regarding the mount points and the direct I/O support.
> 
>     2. Remove the part in this particular test that validates the cache mode for host-path and block PVCs.
> 
> 
> WDYT?

Would exec-ing to `virt-launcher` (and figuring the underlying storage) be possible solution?

**Comment by enp0s3** [comment]
> > @ezrasilvera @rmohr The block volume PV is mounted at **/mnt/local-storage** and the HostPath PV is mounted at **/tmp/hostImages** correct me if I'm wrong but we cannot guarantee that the **/mnt** and **/tmp** are mounted to a FS that supports O_DIRECT. I can suggest 2 possible solutions:
> > ```
> > 1. In our documentation, update the tests section with the relevant information about pre-conditions that should be met regarding the mount points and the direct I/O support.
> > 
> > 2. Remove the part in this particular test that validates the cache mode for host-path and block PVCs.
> > ```
> > 
> > 
> > WDYT?
> 
> Would exec-ing to `virt-launcher` (and figuring the underlying storage) be possible solution?

@xpivarc Thank you for the suggestion! Indeed it might work. In the test suite setup phase, if we will run a VM with these two pre-configured PVCs attached, then read the cache mode from the VMI status and update the PVCs with annotation.

**Comment by ezrasilvera** [comment]
> Would exec-ing to `virt-launcher` (and figuring the underlying storage) be possible solution?

This is a philosophical question :-) of what you actually test. You are going to use an O_DIRECT test for  figuring out what the disk supports, and then test that  the **exact same** O_DIRECT test is working ...

On the other hand because the cache logic your are testing is larger then just the O_DIRECT test I think this might be OK in this case.  
We just need to make sure that we somehow also test our O_DIRECT test logic - taking a **known** FS, and see that our O_DIRECT test works. I guess that the other disk types (ephemeral, etc..) are good enough cover for that. 
So to summarize - I think it's good enough solution for this case

**Comment by enp0s3** [comment]
@ezrasilvera @xpivarc @rmohr I decided to just remove the cache mode testing for block volume and host path PVCs, since we cannot be sure about the underlying FS. I also decided not to add a logic to determine the cache mode in the initialization phase because then I will just test that the cache mode hasn't changed since the initialization, while the aim of the cache mode test is to validate the cache mode that we are 100% sure about them.

**Comment by rmohr** [comment]
@enp0s3 do we still test the cache mode then?

**Comment by enp0s3** [comment]
> @enp0s3 do we still test the cache mode then?

@rmohr Yep. but only for ephemeral disks and host disks, and not for block volume and host-path PVCs

**Review by dhiller — APPROVED** [review]
/lgtm
/approve
/hold in case anyone else wants to have a look

**Comment by ezrasilvera** [comment]
It seems you cover all options of explicitly setting cacheNone and cacheWritethrough as well as testing the auto defaults for theses cases.

/lgtm
